akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/EventSourcedBehaviorWatchSpec.scala

```diff
@@ -61,6 +61,38 @@ class EventSourcedBehaviorWatchSpec
       createTestProbe().expectTerminated(parent)
     }
 
+    "behave as expected if a user's signal handler is side effecting" in {
+      val signalHandler: PartialFunction[(String, Signal), Unit] = {
+        case (_, RecoveryCompleted) =>
+          java.time.Instant.now.getNano
+          Behaviors.same
+      }
+
+      val parent =
+        spawn(Behaviors.setup[Command] { context =>
+          val child = context.spawnAnonymous(Behaviors.receive[Command] { (_, _) =>
+            throw cause
+          })
+
+          context.watch(child)
+
+          EventSourcedBehavior[Command, String, String](nextPid, emptyState = "", commandHandler = (_, cmd) => {
+            child ! cmd
+            Effect.none
+          }, eventHandler = (state, evt) => state + evt).receiveSignal(signalHandler)
+        })
+
+      signalHandler.isDefinedAt(("", RecoveryCompleted)) shouldEqual true
+      signalHandler.isDefinedAt(("", Terminated(parent))) shouldEqual false
+
+      LoggingTestKit.error[TestException].expect {
+        LoggingTestKit.error[DeathPactException].expect {
+          parent ! Fail
+        }
+      }
+      createTestProbe().expectTerminated(parent)
+    }
+
```

Comment:
This is from a chat with @raboof - I am not entirely sure I captured the concern? cc @patriknw

Reply:
A well behaved PartialFunction should work with `isDefinedAt`, or did I miss something? Often avoided for performance reasons but that is not a concern for these signals.

The alternative is to use applyOrElse, right?

We can make the PF total and return `Behaviors.unhandled`. Then we would do some hack in applyOrElse or is there a better way?

```
  def onSignal(state: S, signal: Signal, catchAndLog: Boolean): Boolean = {
    try {
      var handled = true
      signalHandler.applyOrElse((state, signal), { case (_, _) => handled = false })
      handled
    } catch {
      case NonFatal(ex) =>
        if (catchAndLog) {
          log.error(s"Error while processing signal [$signal]: $ex", ex)
          true
        } else {
          if (log.isDebugEnabled)
            log.debug(s"Error while processing signal [$signal]: $ex", ex)
          throw ex
        }
    }
  }
```

and when used:
```
case signal =>
  if (setup.onSignal(state.state, signal, catchAndLog = false))
    Behaviors.same
  else
    Behaviors.unhandled
```

Reply:
> Often avoided for performance reasons but that is not a concern for these signals

Agreed, I was worried over semantics, not performance

> a well behaved `PartialFunction`

I guess indeed the root of the question is 'what is a well behaved `PartialFunction`'.

When you construct a side-effecting `PartialFunction` with `Function.unlift`, they side-effect while evaluating `isDefinedAt` as well:

```
scala> val pf = Function.unlift((a: String) => { println(a); Some(a) })
pf: PartialFunction[String,String] = <function1>scala> pf.isDefinedAt("hoi")
hoi
res4: Boolean = true
```

... so if you'd use such a `PartialFunction` for your signal handler, your signal handler would be ran twice (once while evaluating `isDefinedAt` and once while evaluating `onSignal`)

So the question is if we consider such PF's are defective and unsupported, or that we want to support them. If we want to support them indeed we'd have to do something like that `applyOrElse` hack. In some cases you'd perhaps be able to do `mySignalHandlingPf.orElse(setup.signalHandler)`?

Reply:
ok, let's use `applyOrElse` and return `unhandled`, as in my suggestion above